rend3-routine/src/skinning/mod.rs

```diff
@@ -0,0 +1,272 @@
+            let mut bglb = BindGroupLayoutBuilder::new();
+            bglb.append(
+                wgpu::ShaderStages::COMPUTE,
+                wgpu::BindingType::Buffer {
+                    ty: wgpu::BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: None,
+                },
+                None,
+            );
+            bglb.build(device, Some("Gpu skinning inputs"))
+        };
+
+        let layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
+            label: None,
+            bind_group_layouts: &[&vertex_buffers_bgl, &joint_matrices_bgl, &skinning_inputs_bgl],
+            push_constant_ranges: &[],
+        });
+
+        let module = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
+            label: Some("Gpu skinning compute shader"),
+            source: wgpu::ShaderSource::Wgsl(include_str!("../../shaders/src/skinning.wgsl").into()),
+        });
+
+        let pipeline = device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
+            label: Some("Gpu skinning pipeline"),
+            layout: Some(&layout),
+            module: &module,
+            entry_point: "main",
+        });
+
+        GpuSkinner {
+            vertex_buffers_bgl,
+            joint_matrices_bgl,
+            skinning_inputs_bgl,
+            pipeline,
+        }
+    }
+
+    pub fn execute_pass(
+        &self,
+        device: &Device,
+        encoder: &mut wgpu::CommandEncoder,
+        buffers: &PreSkinningBuffers,
+        mesh_buffers: &MeshBuffers,
+        // The number of inputs in the skinning_inputs buffer
+        skeleton_manager: &SkeletonManager,
+    ) {
+        let vertex_buffers_bg = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&mesh_buffers.vertex_position);
+            bgb.append_buffer(&mesh_buffers.vertex_normal);
+            bgb.append_buffer(&mesh_buffers.vertex_tangent);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_index);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_weight);
+            bgb.build(device, Some("GPU skinning vertex buffers"), &self.vertex_buffers_bgl)
+        };
+
+        let joint_matrices_bg = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&buffers.joint_matrices);
+            bgb.build(device, Some("GPU skinning joint matrices"), &self.joint_matrices_bgl)
+        };
+
+        let skinning_inputs_bgb = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&buffers.gpu_skinning_inputs);
+            bgb.build(device, Some("GPU skinning inputs"), &self.skinning_inputs_bgl)
+        };
+
+        for (i, skel) in skeleton_manager.skeletons().enumerate() {
+            let mut cpass = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
+                label: Some(&format_sso!("GPU Skinning {}", i)),
+            });
```

Comment:
We should lift this cpass outside of the loop -- passes carry some overhead, so we should minimize the count. It also means we can bind the bind groups only once.